### PR TITLE
Cache environment updates in a separate file.

### PR DESCRIPTION
**Pull request by OCamlPro-Henry**
The environment updates are cached in <root>/<switch>/environment.
It is updated whenever the <root>/<switch>/state is updated.

This might be used by external tools (e.g. opam-manager) to read the
environment updates without having to depend on `OpamState`. This might
also permit to add a "costly topological sort" when computing the
environment updates.

### CODE REVIEW

File: src/format/opamFile.ml [hunk 1]
```diff
@@ -437,6 +437,34 @@ module Pinned_legacy = LineFile(struct
...
+
+    type t = env_update list
+
+    let empty = []
+
+    let pp =
+      Pp.lines_set ~empty:[] ~add:OpamStd.List.cons ~fold:List.fold_right @@
+      Pp.map_right
+        (fun (a, (b, (c, d))) -> (a, b, c, d))
+        (fun (a, b, c, d) -> (a, (b, (c, d)))) @@
```
Comment (AltGr) [inline src/format/opamFile.ml]: Isn't this equivalent to:
```
Pp.pp
  (fun (a, (b, (c, d))) -> (a, b, c, d))
  (fun (a, b, c, d) -> (a, (b, (c, d)))) -|
```
?

File: src/format/opamFormat.ml [hunk 2]
```diff
@@ -245,6 +245,20 @@ module Pp = struct
...
+    name_constr = t.name_constr;
+  }
+
+  let map_left f g t = {
+    parse = (fun ~pos x -> t.parse ~pos (g x));
+    print = (fun x -> f (t.print x));
+    name = t.name;
+    name_constr = t.name_constr;
+  }
+
```
Comment (AltGr) [inline src/format/opamFormat.ml]: See above, I think you can define a pp with the two functions and use the normal pipe operator `-|` for the same result.

File: src/state/opamState.ml [hunk 3]
```diff
@@ -2083,7 +2067,30 @@ let env_updates ~opamswitch ?(force_path=false) t =
...
     else []
   in
-  new_path :: (man_path @ switch @ root @ comp_env @ pkg_env)
+  man_path @ root @ comp_env @ pkg_env
+
+let env_updates ~opamswitch ?(force_path=false) t =
+  let update =
+    try
+      OpamFile.Environment.read (OpamPath.Switch.environment t.root t.switch)
+    with _ ->
```
Comment (AltGr) [inline src/state/opamState.ml]: Forbidden! See https://github.com/ocaml/opam/blob/ba3db33874ff1c09b1384c277d8b903748543087/src/core/opamStd.mli#L256-L261

**Comment by AltGr** [comment]
Wow! What a surprise to see patches on code I wrote so recently (and that may still have some rough edges) :)
This is a very good idea. It'll also be easier to debug and the `opam-init/variables.sh` files can be generated in a simpler way. Does this work well with the display of the variables' origin ? (3a231aa2d6df7822e50f77dc9bfabd8d79b564f8, which yields e.g.:
```
[lg@agaric ~/ocamlpro/opam]% opam config env -v
: Updated by compiler 4.02.3;
CAML_LD_LIBRARY_PATH="/home/lg/.opam/4.02.3/lib/stublibs"; export CAML_LD_LIBRARY_PATH;
: Current opam switch man dir;
MANPATH=":/home/lg/.opam/4.02.3/man"; export MANPATH;
: Current opam switch binary dir;
PATH="/home/lg/.opam/4.02.3/bin:/home/lg/bin:/home/lg/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/sbin:/sbin"; export PATH;
```
)

**Comment by AltGr** [comment]
I was actually just writing tests on the `setenv:` field, since I wrote it but didn't test at all yet.

**Comment by AltGr** [comment]
Now that I fixed the `setenv` field I need this so that the cost of `opam config env` doesn't increase too much :)

File: src/format/opamFormat.ml (hunk 2 above)
Comment (OCamlPro-Henry) [inline src/format/opamFormat.ml]: Indeed, fixed.

File: src/state/opamState.ml (hunk 3 above)
Comment (OCamlPro-Henry) [inline src/state/opamState.ml]: Ok! I added a proper test.

File: src/format/opamFile.ml [hunk 4]
```diff
@@ -437,6 +437,38 @@ module Pinned_legacy = LineFile(struct
...
+
+    let pp =
+      Pp.lines_set ~empty:[] ~add:OpamStd.List.cons ~fold:List.fold_right @@
+      Pp.identity ^+
+      Pp.of_module "env_update_op"
+        (module struct
+          type t = env_update_op
+          let to_string = string_of_env_update_op
+          let of_string = env_update_op_of_string
+        end : Pp.STR with type t = env_update_op) ^+
```
Comment (AltGr) [inline src/format/opamFile.ml]:
```
Pp.pp ~name:"env_update_op" string_of_env_update_op env_update_op_of_string
```
would be shorter :)

File: src/client/opamClient.ml [hunk 5]
```diff
@@ -155,6 +155,9 @@ let with_switch_backup command f =
   let t = OpamState.load_state command OpamStateConfig.(!r.current_switch) in
   let file = OpamPath.Switch.backup t.root t.switch in
   OpamFilename.mkdir (OpamPath.Switch.backup_dir t.root t.switch);
+    OpamFile.Environment.write
+      (OpamPath.Switch.environment t.root t.switch)
+      (OpamState.compute_env_updates t);
```
Comment (AltGr) [inline src/client/opamClient.ml]: Why is this needed here ? It's not backup-related and the file should already be there, no ?

Comment (OCamlPro-Henry) [inline src/client/opamClient.ml]: Indeed... removed.

File: src/format/opamFile.ml (hunk 4 above)
Comment (OCamlPro-Henry) [inline src/format/opamFile.ml]: Included. Thanks.

File: src/format/opamFile.ml [hunk 6]
```diff
@@ -437,6 +437,34 @@ module Pinned_legacy = LineFile(struct
...
+
+    type t = env_update list
+
+    let empty = []
+
+    let pp =
+      Pp.lines_set ~empty:[] ~add:OpamStd.List.cons ~fold:List.fold_right @@
+      Pp.identity ^+
+      Pp.pp ~name:"env_update_op"
+        env_update_op_of_string string_of_env_update_op ^+
```
Comment (AltGr) [inline src/format/opamFile.ml]: My  bad, the `of_string` takes an additional `~pos` argument. You can either wrap in `(fun ~pos:_ -> xxx)`, or use `Pp.of_pair "env_update_op" (env_update_op_of_string, string_of_env_update_op)` instead.

Other than that, all good it seems. Thanks!

Comment (OCamlPro-Henry) [inline src/format/opamFile.ml]: Pff... I did not even try to compile... :-( Fixed.
